src/osd/scrubber/scrub_queue.cc

```diff
@@ -0,0 +1,639 @@
+}
+
+
+tl::expected<ScrubPreconds, schedule_result_t>
+ScrubQueue::preconditions_to_scrubbing(
+    const ceph::common::ConfigProxy& config,
+    bool is_recovery_active,
+    utime_t scrub_clock_now)
+{
+  if (auto blocked_pgs = get_blocked_pgs_count(); blocked_pgs > 0) {
+    // some PGs managed by this OSD were blocked by a locked object during
+    // scrub. This means we might not have the resources needed to scrub now.
+    dout(10) << fmt::format(
+		    "{}: PGs are blocked while scrubbing due to locked objects "
+		    "({} PGs)",
+		    __func__, blocked_pgs)
+	     << dendl;
+  }
+
+  // sometimes we just skip the scrubbing
+  if ((rand() / (double)RAND_MAX) < config->osd_scrub_backoff_ratio) {
+    dout(20) << fmt::format(
+		    "{}: lost coin flip, randomly backing off (ratio: {:f})",
+		    __func__, config->osd_scrub_backoff_ratio)
+	     << dendl;
+    return tl::unexpected(schedule_result_t::failure);
+  }
+
+  // fail fast if no resources are available
+  if (!can_inc_scrubs()) {
+    dout(10) << fmt::format("{}: OSD cannot inc scrubs", __func__) << dendl;
+    return tl::unexpected(schedule_result_t::failure);
+  }
+
+  // if there is a PG that is just now trying to reserve scrub replica resources
+  // - we should wait and not initiate a new scrub
+  if (is_reserving_now()) {
+    dout(10) << fmt::format(
+		    "{}: scrub resources reservation in progress", __func__)
+	     << dendl;
+    return tl::unexpected(schedule_result_t::failure);
+  }
+
+  Scrub::ScrubPreconds env_conditions;
+  env_conditions.time_permit = scrub_time_permit();
+  env_conditions.load_is_low = scrub_load_below_threshold();
+  env_conditions.only_deadlined =
+      !env_conditions.time_permit || !env_conditions.load_is_low;
+
+  if (is_recovery_active && !config->osd_scrub_during_recovery) {
+    if (!config->osd_repair_during_recovery) {
+      dout(15) << fmt::format(
+		      "{}: not scheduling scrubs due to active recovery",
+		      __func__)
+	       << dendl;
+      return tl::unexpected(schedule_result_t::failure);
+    }
+
+    dout(10) << fmt::format(
+		    "{}: will only schedule explicitly requested repair due to "
+		    "active recovery",
+		    __func__)
+	     << dendl;
+    env_conditions.allow_requested_repair_only = true;
+  }
+
+  return env_conditions;
+}
+
+/**
+ * the refactored "OSD::sched_all_scrubs()"
+ *
+ * Scans the queue for entries that are "periodic", and messages the PGs
+ * named in those entries to recalculate their scrub scheduling
+ */
+void ScrubQueue::on_config_times_change()
+{
+  std::set<spg_t> to_notify;
+  std::unique_lock l{jobs_lock};
+  for (const auto& e : to_scrub) {
+    if (e.is_valid && e.urgency == urgency_t::periodic_regular) {
+      to_notify.insert(e.pgid);
+    }
+  }
+  l.unlock();
+
+  for (const auto& p : to_notify) {
+    dout(15) << fmt::format("{}: rescheduling {}", __func__, p) << dendl;
+    osd_service.send_sched_recalc_to_pg(p);
```

Comment:
So, it seems like the queue simply needs a way to iterate over its contents, it doesn't actually need to do the notification.